profiler/internal/stackparse/stackparse_test.go

```diff
@@ -0,0 +1,372 @@
+		{
+			Line:  "goroutine 23 [select, 5 minutes]:",
+			WantG: &Goroutine{ID: 23, State: "select", Wait: 5 * time.Minute},
+		},
+		{
+			Line:  "goroutine 42 [select, 5 minutes, locked to thread]:",
+			WantG: &Goroutine{ID: 42, State: "select", Wait: 5 * time.Minute, LockedToThread: true},
+		},
+		{
+			Line:    "goroutine 23 []:",
+			WantErr: "invalid goroutine header",
+		},
+		{
+			Line:    "goroutine ",
+			WantErr: "invalid goroutine header",
+		},
+		{
+			Line:    "goroutine 1 [chan receive]:\n",
+			WantErr: "invalid function call",
+		},
+	},
+
+	funcs: []frameLine{
+		{
+			Line:      "main.main()",
+			WantFrame: &Frame{Func: "main.main"},
+		},
+		{
+			Line:      "runtime.goparkunlock(...)",
+			WantFrame: &Frame{Func: "runtime.goparkunlock"},
+		},
+		{
+			Line:      "net/http.(*persistConn).writeLoop(0xc0001a5c20)",
+			WantFrame: &Frame{Func: "net/http.(*persistConn).writeLoop"},
+		},
+		{
+			Line:    "foo.bar",
+			WantErr: "invalid function call",
+		},
+		{
+			Line:    "foo.bar(",
+			WantErr: "invalid function call",
+		},
+		{
+			Line:    "net/http.(*persistConn).writeLoop(0xc0",
+			WantErr: "invalid function call",
+		},
+		{
+			Line:    "net/http.(*persist",
+			WantErr: "invalid function call",
+		},
+		{
+			Line:    "net/http.*persist)(",
+			WantErr: "invalid function call",
+		},
+		{
+			Line:    "net/http.(*persist))",
+			WantErr: "invalid function call",
+		},
+		{
+			Line:    "net/http.((*persist)",
+			WantErr: "invalid function call",
+		},
+		{
+			Line:    "()",
+			WantErr: "invalid function call",
+		},
+	},
+
+	files: []frameLine{
+		{
+			Line:      "\t/go/src/example.org/example/main.go:231 +0x1187",
+			WantFrame: &Frame{File: "/go/src/example.org/example/main.go", Line: 231},
+		},
+		{
+			Line:      "\t/root/go1.15.6.linux.amd64/src/runtime/proc.go:312",
+			WantFrame: &Frame{File: "/root/go1.15.6.linux.amd64/src/runtime/proc.go", Line: 312},
+		},
+		{
+			Line:    "/root/go1.15.6.linux.amd64/src/runtime/proc.go:312",
+			WantErr: "invalid file:line ref",
+		},
+		{
+			Line:    "",
+			WantErr: "invalid file:line ref",
+		},
+	},
+
+	createdBy: []frameLine{
+		{
+			Line:      "created by net/http.(*Server).Serve",
+			WantFrame: &Frame{Func: "net/http.(*Server).Serve"},
+		},
+		{
+			Line:      "created by github.com/example.org/example/k8s.io/klog.init.0",
+			WantFrame: &Frame{Func: "github.com/example.org/example/k8s.io/klog.init.0"},
+		},
+	},
+}
+
+func BenchmarkParse(b *testing.B) {
+	data, err := ioutil.ReadFile(filepath.Join("test-fixtures", "waitsince.txt"))
+	require.NoError(b, err)
+
+	b.ResetTimer()
+	b.ReportAllocs()
+
+	start := time.Now()
+	parsedBytes := 0
+	for i := 0; i < b.N; i++ {
+		parsedBytes += len(data)
+		gs, err := Parse(bytes.NewReader(data))
+		if err != nil {
+			b.Fatal(err)
+		} else if l := len(gs); l != 9 {
+			b.Fatal(l)
+		}
+	}
+
+	mbPerSec := float64(parsedBytes) / time.Since(start).Seconds() / 1024 / 1024
+	b.ReportMetric(mbPerSec, "MB/s")
+}
```

Comment:
The benchmark reports ~300 MB/s on my machine which is pretty good. I also used cpu/heap profiling to pick some low-hanging fruits to make things faster. At this point there is probably no ROI in optimizing the parser further.

Also worth knowing: The stack dump is [limited to 64MB](https://go.googlesource.com/go/+/go1.15.6/src/runtime/pprof/pprof.go#698), so we'd use a maximum of ~200ms of CPU time during each continuous profiling iteration.

Reply:
👍